internal/service/s3/bucket_acl.go

```diff
@@ -0,0 +1,487 @@
+		owner.DisplayName = aws.String(v)
+	}
+
+	if v, ok := tfMap["id"].(string); ok && v != "" {
+		owner.ID = aws.String(v)
+	}
+
+	return owner
+}
+
+func flattenBucketAclAccessControlPolicy(output *s3.GetBucketAclOutput) []interface{} {
+	if output == nil {
+		return []interface{}{}
+	}
+
+	m := make(map[string]interface{})
+
+	if len(output.Grants) > 0 {
+		m["grant"] = flattenBucketAclAccessControlPolicyGrants(output.Grants)
+	}
+
+	if output.Owner != nil {
+		m["owner"] = flattenBucketAclAccessControlPolicyOwner(output.Owner)
+	}
+
+	return []interface{}{m}
+}
+
+func flattenBucketAclAccessControlPolicyGrants(grants []*s3.Grant) []interface{} {
+	var results []interface{}
+
+	for _, grant := range grants {
+		if grant == nil {
+			continue
+		}
+
+		m := make(map[string]interface{})
+
+		if grant.Grantee != nil {
+			m["grantee"] = flattenBucketAclAccessControlPolicyGrantsGrantee(grant.Grantee)
+		}
+
+		if grant.Permission != nil {
+			m["permission"] = aws.StringValue(grant.Permission)
+		}
+
+		results = append(results, m)
+	}
+
+	return results
+}
+
+func flattenBucketAclAccessControlPolicyGrantsGrantee(grantee *s3.Grantee) []interface{} {
+	if grantee == nil {
+		return []interface{}{}
+	}
+
+	m := make(map[string]interface{})
+
+	if grantee.DisplayName != nil {
+		m["display_name"] = aws.StringValue(grantee.DisplayName)
+	}
+
+	if grantee.EmailAddress != nil {
+		m["email_address"] = aws.StringValue(grantee.EmailAddress)
+	}
+
+	if grantee.ID != nil {
+		m["id"] = aws.StringValue(grantee.ID)
+	}
+
+	if grantee.Type != nil {
+		m["type"] = aws.StringValue(grantee.Type)
+	}
+
+	if grantee.URI != nil {
+		m["uri"] = aws.StringValue(grantee.URI)
+	}
+
+	return []interface{}{m}
+}
+
+func flattenBucketAclAccessControlPolicyOwner(owner *s3.Owner) []interface{} {
+	if owner == nil {
+		return []interface{}{}
+	}
+
+	m := make(map[string]interface{})
+
+	if owner.DisplayName != nil {
+		m["display_name"] = aws.StringValue(owner.DisplayName)
+	}
+
+	if owner.ID != nil {
+		m["id"] = aws.StringValue(owner.ID)
+	}
+
+	return []interface{}{m}
+}
+
+// BucketACLCreateResourceID is a method for creating an ID string
+// with the bucket name and optional accountID and/or ACL.
+func BucketACLCreateResourceID(bucket, expectedBucketOwner, acl string) string {
+	if expectedBucketOwner == "" {
+		if acl == "" {
+			return bucket
+		}
+		return strings.Join([]string{bucket, acl}, BucketAclSeparator)
+	}
+
+	if acl == "" {
+		return strings.Join([]string{bucket, expectedBucketOwner}, BucketAndExpectedBucketOwnerSeparator)
+	}
+
+	parts := []string{bucket, expectedBucketOwner}
+	id := strings.Join([]string{strings.Join(parts, BucketAndExpectedBucketOwnerSeparator), acl}, BucketAclSeparator)
+
+	return id
+}
+
+// BucketACLParseResourceID is a method for parsing the ID string
+// for the bucket name, accountID, and ACL if provided.
+func BucketACLParseResourceID(id string) (string, string, string, error) {
```

Comment:
maybe this is too much just to store the optional "acl" ?

Reply:
Maybe a regex would be more straightforward for parsing